README.md
[![Build Status](https://travis-ci.org/laurieodgers/koa-api-framework.svg?branch=master)](https://travis-ci.org/laurieodgers/koa-api-framework)

# koa-api-framework

**Migrating to koa-api-framework 2 / Koa 2**

- The framework has changed to match the new promise-based middleware signature of koa 2. See the [koa 2.x readme](https://github.com/koajs/koa/tree/2.0.0-alpha.3) for more information or see the example for an example on how to use async/await.

## Description

An API framework using Koa.

koa-api-framework builds an API framework from a RAML 0.8 file, with automatic payload validation and JWT decoding for your endpoints.

## Features
### An API From RAML
Your API will be constructed based on RAML. Specify the location in ./raml/api.raml or through the `raml` argument to the constructor.

### Simple Request Validation
Automatic validation of incoming JSON data is performed if you specify the JSON schema within your RAML file, removing the need for validation within each individual endpoint.

### Automatic JWT decoding
Set RAML traits on your endpoints to indicate that a JWT is required. Any incoming `Authorization: Bearer (jwt)` headers for these endpoints will automatically decode the given JWT and place it within `this.token`.

## Directory Structure
The controller path may be passed in with `controllerPath` to the constructor. The directory structure underneath must match the part of the path after `apiBase` with respect to the URL.

See /example for a complete example on how to construct your API.

## Returning Data
- `this.data` within your controllers allows you to return data to the client in the prescribed data structure. If you wish to override this with your own data structure, then assign your response body to `this.body`.
- Any errors thrown within your controllers will be caught by the framework in order to always present JSON to the client. To throw an error and return a HTTP status code and message to the client, use `throw new Error("418:I'm a teapot!");` where 418 is the status code, and "I'm a teapot!" is the error message to the user.

## Example
A complete example can be found under `/example`.

## Contributing
Contributions are always welcome; if you fix a bug or implement some extra functionality please issue a PR back to https://github.com/laurieodgers/koa-api-framework

## Installation
Using npm:
```
npm install --save koa-api-framework
```

In Node.js:
```
var KoaApiFramework = require('koa-api-framework');

// set up the framework
var framework = new KoaApiFramework({
    // disable extra debug logging
    debug: false,
    // enable SSL/TLS
    tls: true,
    // the ssl/tls private key
    tlsKeyPath: '/path/to/tls/key/server.key',
    // the ssl/tls certificate path
    tlsCertPath: '/path/to/tls/cert/server.crt',
    // HTTP(s) port
    port: 443,
    // pass in a path to a raml file
    // DEFAULT: './raml/api.raml'
    raml: './raml/api.raml',
    // The base URL of this API
    apiBase: 'https://api.foo.com/v2',
    // pass in any models for your constructors to consume
    models: {
        auth: auth,
        person: person,
        group: group
    },
    // specify where the controllers are in the filesystem
    // DEFAULT: ./controllers
    controllerPath: '/ctrl',
    // koa-api-framework will automatically decode Authorization: Bearer [jwt] for you
    // for endpoints which are specified in 'authTraits'
    jwtSecret: 'qwertyuiopasdfghjklzxcvbnm123456',
    // set which traits will decode the JWT
    // DEFAULT: ['authenticated']
    authTraits: ['authenticated', 'administrator', 'myAuthenticatedTrait']
});
```


CHANGELOG.md
# 2.1.0
- Completely removed co code
- Added check for exp to ensure that the token has not expired

# 2.0.0
- Fixed deprecation messages. Raised to major version due to lack of backwards compatibility. (closes #4)

# 1.1.1
- Added checks to ensure TLS cert and key exist (closes #7)
- Fixed error in controllers not displaying to console (closes #8)
- Added ability to override default json structure by setting this.body directly

# 1.1.0
- Changed required field within JWT to 'sub' from 'userId'
- Added ability to add your own RAML trait names that should be authenticated by JWT
- Fixed: trailing slash on apiBase to constructor no longer causes a double slash. Fixes #5
- Added ability to define where the controllers are in the filesystem
- Added travisci tests
- Added checks for content-type when content-length is specified


tests/test.sh
#!/bin/bash

set -e

jwtWithNoSub="eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpYXQiOjE1MTYyNzMwNjEuNTA0LCJ0ZXN0IjoidGVzdCJ9.tYQIrlVX7A4jaib852v6RauFlKiokcXAanfqHKZogMQ"
jwtWithSub="eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJ0ZXN0IiwiaWF0IjoxNTE2Mjc4NTExLjAxNCwidGVzdCI6InRlc3QifQ.3CBFs7JXh70C6qC8FsSQK2jYwsEYMJbDp1wgp8ltr3E"
jwtNotOurs="eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiYWRtaW4iOnRydWV9.TJVA95OrM7E2cBab30RMHrHDcEfxjoYZgeFONFh7HgQ"

expiredJWT=$( node tests/createjwt.js -100 )
nonExpiredJWT=$( node tests/createjwt.js 100 )


node tests/server.js &

serverPid=$!

sleep 5

function exitWithError() {
    kill ${serverPid}
    exit 1
}

# test endpoints
echo "GET http://localhost:8080/v2/user/whoami"
echo "1. Valid JWT"
output=$( curl -s -H "Authorization: Bearer ${jwtWithSub}" -X "GET" http://localhost:8080/v2/user/whoami )
echo "${output}"
if [[ "${output}" != '{"statusCode":200,"message":"","data":{"userId":"1"}}' ]]; then
    echo "FAILED"
    exitWithError
fi
echo ""
echo "2. No Authorization Header"
output=$( curl -s -X "GET" http://localhost:8080/v2/user/whoami )
echo "${output}"
if [[ "${output}" != '{"status":401,"message":"Unauthorized","data":{}}' ]]; then
    echo "FAILED"
    exitWithError
fi
echo ""
echo "3. No JWT"
output=$( curl -s -H "Authorization: Bearer " -X "GET" http://localhost:8080/v2/user/whoami )
echo "${output}"
if [[ "${output}" != '{"status":400,"message":"Invalid JWT","data":{}}' ]]; then
    echo "FAILED"
    exitWithError
fi
echo ""
echo "4. Invalid JWT"
output=$( curl -s -H "Authorization: Bearer asdfasdfasdfasdf" -X "GET" http://localhost:8080/v2/user/whoami )
echo "${output}"
if [[ "${output}" != '{"status":400,"message":"Invalid JWT","data":{}}' ]]; then
    echo "FAILED"
    exitWithError
fi
echo ""
echo "5. Valid JWT not ours"
output=$( curl -s -H "Authorization: Bearer ${jwtNotOurs}" -X "GET" http://localhost:8080/v2/user/whoami )
echo "${output}"
if [[ "${output}" != '{"status":400,"message":"Invalid JWT","data":{}}' ]]; then
    echo "FAILED"
    exitWithError
fi
echo ""
echo "6. Valid JWT but no sub"
output=$( curl -s -H "Authorization: Bearer ${jwtWithNoSub}" -X "GET" http://localhost:8080/v2/user/whoami )
echo "${output}"
if [[ "${output}" != '{"status":400,"message":"Invalid JWT - Missing Subject","data":{}}' ]]; then
    echo "FAILED"
    exitWithError
fi
echo ""
echo "7. Expired JWT"
output=$( curl -s -H "Authorization: Bearer ${expiredJWT}" -X "GET" http://localhost:8080/v2/user/whoami )
echo "${output}"
if [[ "${output}" != '{"status":401,"message":"Invalid JWT - Expired","data":{}}' ]]; then
    echo "FAILED"
    exitWithError
fi
echo ""
echo "7. Non Expired JWT"
output=$( curl -s -H "Authorization: Bearer ${nonExpiredJWT}" -X "GET" http://localhost:8080/v2/user/whoami )
echo "${output}"
if [[ "${output}" != '{"statusCode":200,"message":"","data":{"userId":"1"}}' ]]; then
    echo "FAILED"
    exitWithError
fi

echo ""
echo "----"
echo ""

echo "POST http://localhost:8080/v2/person"
echo "1. Valid POST data"
output=$( curl -s -H "Authorization: Bearer ${jwtWithSub}" -H "Content-type: application/json" -X "POST" -d '{"name":"test"}' http://localhost:8080/v2/person )
echo "${output}"
if [[ "${output}" != '{"statusCode":200,"message":"","data":{"message":"POST person endpoint"}}' ]]; then
    echo "FAILED"
    exitWithError
fi
echo ""

echo "2. Invalid Content Type"
output=$( curl -s -H "Authorization: Bearer ${jwtWithSub}" -H "Content-type: text/plain" -X "POST" -d '{"name":"test"}' http://localhost:8080/v2/person )
echo "${output}"
if [[ "${output}" != "{\"status\":415,\"message\":\"Content-type 'text/plain' not supported\",\"data\":{}}" ]]; then
    echo "FAILED"
    exitWithError
fi
echo ""

echo "3. No Content Type"
output=$( curl -s -H "Authorization: Bearer ${jwtWithSub}" -H "Content-type: " -X "POST" -d '{"name":"test"}' http://localhost:8080/v2/person )
echo "${output}"
if [[ "${output}" != '{"status":415,"message":"Content-type not specified","data":{}}' ]]; then
    echo "FAILED"
    exitWithError
fi
echo ""

echo ""
echo "----"
echo ""

echo "GET http://localhost:8080/v2/person/list"
echo "1. Valid GET request"
output=$( curl -s -H "Authorization: Bearer ${jwtWithSub}" -X "GET" http://localhost:8080/v2/person/list )
echo "${output}"
if [[ "${output}" != '{"statusCode":200,"message":"","data":{"people":[{"name":"Jane"},{"name":"Bob"}]}}' ]]; then
    echo "FAILED"
    exitWithError
fi
echo ""

echo "2. Invalid GET request"
output=$( curl -s -H "Authorization: Bearer ${jwtWithSub}" -X "GET" http://localhost:8080/v2/nonexistent )
echo "${output}"
if [[ "${output}" != '{"status":404,"message":"Endpoint not found","data":{}}' ]]; then
    echo "FAILED"
    exitWithError
fi
echo ""

echo ""
echo "----"
echo ""

echo "GET http://localhost:8080/v2/errors/500"
echo "1. Check 500 error"
output=$( curl -s -X "GET" http://localhost:8080/v2/errors/500 )
statusCode=$( curl -s -o /dev/null -w "%{http_code}" http://localhost:8080/v2/errors/500 )
echo "${output}"
if [[ "${output}" != '{"status":500,"message":"500 internal server error","data":{}}'  || ${statusCode} -ne 500 ]]; then
    echo "FAILED"
    echo "Status Code: ${statusCode}"
    echo "Output: ${output}"
    exitWithError
fi
echo ""
echo "2. Check user error and message"
output=$( curl -s -X "GET" http://localhost:8080/v2/errors/418 )
statusCode=$( curl -s -o /dev/null -w "%{http_code}" http://localhost:8080/v2/errors/418 )
echo "${output}"
if [[ "${output}" != '{"status":418,"message":"I am a teapot","data":{}}' || ${statusCode} -ne 418 ]]; then
    echo "FAILED"
    exitWithError
fi
echo ""

echo ""
echo "----"
echo ""
url=http://localhost:8080/v2/body/test1""
echo "GET ${url}"
echo "1. Check body override"
output=$( curl -s -X "GET" ${url} )
statusCode=$( curl -s -o /dev/null -w "%{http_code}" ${url} )
echo "${output}"
if [[ "${output}" != '{"hello":"world!"}'  || ${statusCode} -ne 200 ]]; then
    echo "FAILED"
    echo "Status Code: ${statusCode}"
    echo "Output: ${output}"
    exitWithError
fi
echo ""

# kill the server
kill ${serverPid}

echo ""
echo "----"
echo ""
echo "Tests succeeded"


tests/server.js
'use strict';

(async () => {
    var KoaApiFramework = require('../index');
    
    // set up the framework
    var framework = new KoaApiFramework({
        debug: true,
        tls: false,
        port: 8080,
        apiBase: 'https://api.foo.com/v2/',
        raml: './tests/raml/api_specification.raml',
        controllerPath: '/tests/controllers',
        models: {},
        jwtSecret: 'qwertyuiopasdfghjklzxcvbnm123456',
        authTraits: ['authenticated', 'administrator']
    });

    // run the framework
    await framework.start();
})().catch(e => {
    console.error(e);
});;


tests/createjwt.js
var jwt = require('jwt-simple');
var moment = require('moment');

var expTime = process.argv[2];

var exp = moment().add(expTime, 'seconds');

var payload = {
    sub: 'test',
    iat: new Date() / 1000,
    test: 'test',
    exp: exp.toISOString()
}

console.log(jwt.encode(payload, 'qwertyuiopasdfghjklzxcvbnm123456'));


index.js
'use strict';
var _ = require("lodash");
var koa = require('koa');
var koaRouter = require('koa-router');
var http = require('http');
var https = require('https');
var fs = require('fs');
var bodyParser = require("koa-bodyparser");
var cors = require('koa2-cors');
var ramlParser = require('raml-parser');
var jwt = require('jwt-simple');
var JSONValidator = require('jsonschema').Validator;
var Moment = require('moment');

var v = new JSONValidator();

module.exports = class KoaApiFramework {
    constructor(obj) {
        var self = this;

        self.database = obj.database;
        self.models = obj.models;
        self.jwtSecret = obj.jwtSecret;
        self.pathBase = processUri(obj.apiBase) || '';
        self.raml = obj.raml || './raml/api.raml';
        this.authTraits = obj.authTraits || ['authenticated'];

        if (obj.controllerPath) {
            // strip trailing slash
            this.controllerPath = obj.controllerPath.replace(/\/+$/,'');
        } else {
            this.controllerPath = '/controllers';
        }

        obj = _.pick(obj, [
            'tls',
            'tlsKeyPath',
            'tlsCertPath',
            'port',
            'debug'
        ]);
        self.options = obj;

        self.router = new koaRouter();
    }

    async start() {
        var self = this;

        // sanity checks
        if (self.tlsKeyPath && !fs.existsSync(self.tlsKeyPath)) {
            console.error("TLS key does not exist: " + self.tlsKeyPath);
            process.exit(1);
        }
        if (self.tlsCertPath && !fs.existsSync(self.tlsCertPath)) {
            console.error("TLS certificate does not exist: " + self.tlsCertPath);
            process.exit(1);
        }

        var app = new koa();
        app
            .use(cors())
            .use(bodyParser());

        app.use(async (ctx, next) => {
            ctx.db = self.database;
            ctx.models = self.models;
            ctx.jwtSecret = self.jwtSecret;
            ctx.authTraits = self.authTraits;
            ctx.controllerPath = self.controllerPath;

            // set up data so we can use it downstream
            ctx.data = {};

            var urlArray = ctx.request.url.substr(1).split('/');
            ctx.ramlRef = ctx.request.url;

            // strip the queryString
            var searchString = ctx.request.url.split('?')[0];

            // search for this endpoint in our list for easy validation
            for (var i = 0; i < self.endpoints.length; i++) {
                if (
                    (self.endpoints[i].method.toLowerCase() == ctx.request.method.toLowerCase()) &&
                    (searchString.toLowerCase().match(self.endpoints[i].regexPath))
                ) {
                    ctx.endpoint = self.endpoints[i];
                }
            }

            if (!ctx.endpoint) {
                returnError(ctx, 404, "Endpoint not found");
                return;
            }

            var contentType;

            // validate the content type
            if (ctx.request.header['content-type']) {
                var contentTypeArray = ctx.request.header['content-type'].toLowerCase().split(';');
                contentType = contentTypeArray[0];

                // if we don't accept this content type then reject it
                if (contentType != 'application/json') {
                    returnError(ctx, 415, "Content-type '" + contentType + "' not supported");
                    return;
                }
            } else if (ctx.request.header['content-length']) {  // content length is specified but content type is not
                returnError(ctx, 415, "Content-type not specified");
                return;
            }

            // validate the request body
            if (ctx.endpoint.requestSchema) {
                if (!ctx.request.body) {
                    throw new Error("400:No data received");
                }

                var validation = v.validate(ctx.request.body, ctx.endpoint.requestSchema);

                if (validation.errors.length > 0) {
                    throw new Error("400:" + validation.errors.join(", "));
                }
            }


            // process JWT for JWT endpoints
            if (ctx.endpoint.traits && ctx.authTraits.some(r=> ctx.endpoint.traits.indexOf(r) >= 0)> -1) {

                // make sure it has an authorization header
                if (!ctx.request.header.authorization) {
                    returnError(ctx, 401, "Unauthorized");
                    return;
                }

                // remove Bearer from the start of the string
                var jwtString = ctx.request.header.authorization.replace(/^([Bb][Ee][Aa][Rr][Ee][Rr]\s)/,"");

                try {
                    // process jwt and store for later use
                    ctx.token = jwt.decode(jwtString, self.jwtSecret);
                } catch (err) {
                    // not a valid JWT
                    returnError(ctx, 400, "Invalid JWT");
                    return;
                }

                // make sure the JWT includes a subject
                if (!ctx.token.sub) {
                    // not a valid JWT
                    returnError(ctx, 400, "Invalid JWT - Missing Subject");
                    return;
                }

                // make sure the JWT hasnt expired
                if (ctx.token.exp) {
                    var now = Moment();
                    var jwtExp = Moment(ctx.token.exp);

                    if (jwtExp < now) {
                        returnError(ctx, 401, "Invalid JWT - Expired");
                        return;
                    }
                }
            }

            try {

                await next();

                // TODO: validate the response body
                /*
                if (ctx.endpoint.responseSchema) {
                    var validation = v.validate(ctx.request.body, ctx.endpoint.responseSchema);

                    if (validation.errors.length > 0) {
                        throw new Error("400:" + validation.errors.join(", "));
                    }
                }*/
            } catch(err) {
                returnError(ctx, 500, "An internal error occurred", err);
                return;
            }

            // everything worked so return status 200 and the data
            ctx.status = 200;

            // if the user didn't define this.body then set up a structure for them
            if (!ctx.body) {
                ctx.body = {
                    statusCode: 200,
                    message: '',
                    data: ctx.data
                }
            }
        });

        this.endpointValidation = {
            request: {},
            response: {}
        };

        // parse the raml and set up the API structure
        ramlParser.loadFile(self.raml).then(function(raml) {
            console.log('------');
            console.log(raml.title);
            console.log('------');

            // if the pathBase wasnt provided then retrieve it from RAML
            if (!self.pathBase) {
                self.pathBase = processUri(raml.baseUri);
            }


            self.endpoints = [];

            try {
                processEndpoints(raml.resources, '', '');
            } catch (err) {
                console.error(err);
            }

            app
                .use(self.router.routes())
                .use(self.router.allowedMethods());

            var callback = app.callback();

            var server;

            if (!self.options.tls) {
                server = http.createServer(callback);
            } else {

                try {
                    var key = fs.readFileSync(self.options.tlsKeyPath);
                    var cert = fs.readFileSync(self.options.tlsCertPath);
                } catch (err) {
                    console.error(err);
                }

                server = https.createServer({
                    key: key,
                    cert: cert
                }, callback);
            }
            server.timeout = 180*60 * 1000;

            server.listen(self.options.port);

            console.log("Listening on port " + self.options.port);
        }, function(error) {
            console.error("Error parsing RAML: " + error);
            process.exit(1);
        });

        // sets up ctx to return an error message
        function returnError(ctx, statusCode, errorMessage, caughtErr) {
            var error;
            var statusCode;

            // err.message is in the format "statusCode:error message"
            if (caughtErr) {
                // check if the message contains a : indicating the user generated it
                if (caughtErr.message.indexOf(':') == 3) {
                    error = caughtErr.message.split(':');
                    statusCode = parseInt(error[0]);
                    errorMessage = error[1];
                } else {
                    // this wasnt a user generated error
                    if (self.options.debug) {
                        errorMessage = caughtErr.message;
                    } else {
                        errorMessage = errorMessage;
                    }

                    // display to logs
                    console.error(caughtErr);
                }
            }

            // handle errors other than our own
            if (isNaN(statusCode)) {
                statusCode = 500;
            }

            ctx.body = {
                status: statusCode,
                message: errorMessage,
                data: {}
            };
            ctx.status = statusCode;

            return;
        }

        // set up routes recursively
        function processEndpoints(resources, parentPath, parentRegex) {
            for (var i = 0; i < resources.length; i++) {
                var routerPath;
                var thisPath;
                var regexPath;

                // handle variables in the uri
                if (resources[i].relativeUri.startsWith('/{') && resources[i].relativeUri.endsWith('}')) {
                    var param = resources[i].relativeUri.replace(/\/{/gi, '');
                    var param = param.replace(/}/gi, '');

                    var thisPath = parentPath;
                    var routerPath = self.pathBase + parentPath + '/:' + param;
                    var regexPath = parentRegex + '/(.+)';
                } else {
                    var thisPath = parentPath + resources[i].relativeUri;
                    var routerPath = self.pathBase + thisPath;
                    var regexPath = parentRegex + resources[i].relativeUri;
                }

                // check for existence of methods (get/post/put) on this endpoint
                if (resources[i].methods) {

                    // get the file and register the routes
                    var controllerAbsolutePath = process.cwd() + self.controllerPath + '/' + thisPath + '/index';
                    try {
                        var endpoints = require(controllerAbsolutePath);
                    } catch (err) {
                        if (err.code == 'MODULE_NOT_FOUND') {
                            console.error("ERROR: Controller not found at " + controllerAbsolutePath);
                            process.exit(1);
                        } else {
                            throw err;
                        }
                    }

                    for (var j = 0; j < resources[i].methods.length; j++) {
                        var method = resources[i].methods[j];
                        var key = resources[i].methods[j].method;

                        if (endpoints.hasOwnProperty(key)) {
                            // print a message showing this endpoint
                            console.log(key.toUpperCase() + ' ' + routerPath);

                            var requestSchema = null;
                            var responseSchema = null;
                            if (method && method.body && method.body['application/json'] && method.body['application/json'].schema) {
                                requestSchema = JSON.parse(method.body['application/json'].schema);
                            }
                            if (method && method.responses && method.responses['200'] && method.responses['200'].body && method.responses['200'].body['application/json'] && method.responses['200'].body['application/json'].schema) {
                                responseSchema = JSON.parse(method.responses['200'].body['application/json'].schema);
                            }

                            // save endpoint data for processing incoming requests
                            self.endpoints.push({
                                path: thisPath,
                                regexPath: '^' + self.pathBase + regexPath + '$',
                                method: key,
                                traits: resources[i].methods[j].is,
                                requestSchema: requestSchema,
                                responseSchema: responseSchema
                            });

                            // register the endpoint
                            self.router[key](routerPath, endpoints[key]);
                        }
                    }
                }

                // process children if they exist
                if (resources[i].resources) {
                    processEndpoints(resources[i].resources, thisPath, regexPath);
                }
            }
        }
    }
}

// split out the base path from the given api url
function processUri(uri) {
    if (!uri) {
        return '';
    }
    // remove any trailing slashes
    uri = uri.replace(/\/+$/,'');

    var pathBase = uri.split('://')[1];
    pathBase = pathBase.split('/');
    delete pathBase[0];

    return pathBase.join('/');
}
